pkg/etcdcli/etcdcli.go

```diff
@@ -329,9 +317,63 @@ func (g *etcdClientGetter) UnhealthyMembers() ([]*etcdserverpb.Member, error) {
 		g.eventRecorder.Warningf("UnhealthyEtcdMember", "unhealthy members: %v", strings.Join(unhealthyMemberNames, ","))
 	}
 
+	_, unhealthyMembers := etcdMemberHealth.MemberStatus()
+
 	return unhealthyMembers, nil
 }
 
+func (g *etcdClientGetter) MemberHealth(etcdMembers []*etcdserverpb.Member) (*memberHealth, error) {
+	memberHealth, err := g.getMemberHealth(etcdMembers)
+	if err != nil {
+		return nil, err
+	}
+	return memberHealth, nil
+}
+
+func (g *etcdClientGetter) getMemberHealth(etcdMembers []*etcdserverpb.Member) (*memberHealth, error) {
+	var wg sync.WaitGroup
+	memberHealth := &memberHealth{}
+	hch := make(chan healthCheck, len(etcdMembers))
+	for _, member := range etcdMembers {
+		if len(member.ClientURLs) == 0 {
+			memberHealth.Check = append(memberHealth.Check, healthCheck{Member: member, Health: false, Started: false})
+			continue
+		}
+		wg.Add(1)
+		go func(member *etcdserverpb.Member) {
+			defer wg.Done()
+			cli, err := getEtcdClient([]string{member.ClientURLs[0]})
```

Comment:
Might be worth a comment here that we're intentionally making a new client per request for simplicity/bug avoidance and that the cost has been measured and deemed insignificant.

In fact, I would extend the line of reasoning to the use of cached clients/connections generally here given the complexity already present. I don't have any numbers to justify the use of a cached client here or anywhere else, but I also don't want to suggest the caching was introduced for no reason. Maybe with other problems addressed (e.g. less noisy steady state) the problem changes.

Reply:
I added a comment to this regard and we can continue the conversation.